### PR TITLE
canary longform inference script with timestamps option

### PR DESCRIPTION
**Pull request by krishnacpuvvada**
> [!IMPORTANT]  
> The `Update branch` button must only be pressed in very rare occassions.
> An outdated branch is never blocking the merge of a PR.
> Please reach out to the automation team before pressing that button.

# What does this PR do ?

Adds support for timestamps in long-form inference script for AED models.

**Collection**: [ASR]

# Changelog

- Add specific line by line info of high level changes in this PR.

# Usage

- You can potentially add a usage example below

```python
python speech_to_text_aed_chunked_infer.py \
    model_path=null \
    pretrained_name="nvidia/canary-1b-flash" \
    audio_dir="<(optional) path to folder of audio files>" \
    dataset_manifest="<(optional) path to manifest>" \
    output_filename="<(optional) specify output filename>" \
    chunk_len_in_secs=10.0 \
    batch_size=16 \
    decoding.beam.beam_size=1 \
    timestamps=True
```
results for tedlium-3 long-form inference using above script

canary-1b-flash:
chunk_len_in_secs=40.0, compute_timestamps=False : 4.40 (no PnC)
chunk_len_in_secs=10.0, compute_timestamps=True : 5.14 (no PnC)

canary-180m-flash:
chunk_len_in_secs=40.0, compute_timestamps=False : 4.37 (no PnC)
chunk_len_in_secs=10.0, compute_timestamps=True : 5.69 (no PnC)

# GitHub Actions CI

The Jenkins CI system has been replaced by GitHub Actions self-hosted runners.

The GitHub Actions CI will run automatically when the "Run CICD" label is added to the PR.
To re-run CI remove and add the label again.
To run CI on an untrusted fork, a NeMo user with write access must first click "Approve and run".

# Before your PR is "Ready for review"

**Pre checks**:

- [X] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [X] Did you write any new necessary tests?
- [X] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:

- [X] New Feature
- [ ] Bugfix
- [X] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed.
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information

- Related to # (issue)

### CODE REVIEW

File: examples/asr/asr_chunked_inference/aed/speech_to_text_aed_chunked_infer.py [hunk 1]
```diff
@@ -33,13 +33,16 @@
...
+    pretrained_name="nvidia/canary-1b-flash" \
     audio_dir="<(optional) path to folder of audio files>" \
     dataset_manifest="<(optional) path to manifest>" \
     output_filename="<(optional) specify output filename>" \
     chunk_len_in_secs=40.0 \
     batch_size=16 \
-    decoding.beam.beam_size=5
+    decoding.beam.beam_size=1
+
+To return word and segment level timestamps, add `compute_timestamps=True` to the above command,
```
Comment (nithinraok) [inline examples/asr/asr_chunked_inference/aed/speech_to_text_aed_chunked_infer.py:44]: Can you update `compute_timestamps=True` to `timestamps=True` we updated this in transcribe_speech.py script as well.

File: nemo/collections/asr/parts/utils/streaming_utils.py [hunk 2]
```diff
@@ -1672,12 +1700,102 @@ def transcribe(
         """
         self.infer_logits(keep_logits)
 
-        hypothesis = " ".join([h.text for h in self.all_preds])
+        # join hypotheses 
+        hypothesis = self._join_hypotheses(self.all_preds)
```
Comment (nithinraok) [inline nemo/collections/asr/parts/utils/streaming_utils.py:1704]: instead of this `predictions = self.asr_model.predict_step(batch_input, has_processed_signal=True)` why not use `process_aed_timestamp_outputs` with an option to pass offset to it? basically combination of `_transcribe_forward` + `_transcribe_output_processing`

File: nemo/collections/asr/parts/utils/timestamp_utils.py [hunk 3]
```diff
@@ -78,6 +78,7 @@ def segments_offset_to_time(segments, window_stride, subsampling_factor):
 
     for idx, hyp in enumerate(outputs):
         timestamp, text = extract_words_with_timestamps(hyp.text, subsampling_factor, window_stride)
+        text = text.strip()     # remove leading and trailing whitespaces - training data artifact
```
Comment (nithinraok) [inline nemo/collections/asr/parts/utils/timestamp_utils.py:81]: instead of here, do this at the beginning of the `extract_words_with_timestamps` function.

File: .github/workflows/cicd-main.yml [hunk 4]
```diff
@@ -649,6 +649,52 @@ jobs:
...
+  # L2: Longform without TimeStamps from Audio Dir
+  L2_Longform_Speech_Transcription_Canary_Chunked_Infer_from_Audio_Dir:
+    needs: [pre-flight, cicd-test-container-build]
+    uses: ./.github/workflows/_test_template.yml
+    if: contains(fromJSON(needs.pre-flight.outputs.test_to_run), 'L2_Longform_Speech_Transcription_Canary_Chunked_Infer_from_Audio_Dir')
+    with:
+      RUNNER: self-hosted-azure
+      SCRIPT: |
+        python examples/asr/asr_chunked_inference/aed/speech_to_text_aed_chunked_infer.py \
+        pretrained_name=nvidia/canary-1b-flash \
```
Comment (nithinraok) [inline .github/workflows/cicd-main.yml:661]: copy flash model to ci-cd testdata folder and avoid downloading each time.

Comment (nithinraok) [reply .github/workflows/cicd-main.yml:661]: Mandatory~

**Review by nithinraok — CHANGES_REQUESTED** [review]

File: examples/asr/asr_chunked_inference/aed/speech_to_text_aed_chunked_infer.py (hunk 1 above)
Comment (krishnacpuvvada) [reply examples/asr/asr_chunked_inference/aed/speech_to_text_aed_chunked_infer.py:44]: updated.

File: .github/workflows/cicd-main.yml (hunk 4 above)
Comment (krishnacpuvvada) [reply .github/workflows/cicd-main.yml:661]: done.

File: .github/workflows/cicd-main.yml [hunk 5]
```diff
@@ -649,6 +649,52 @@ jobs:
...
+        audio_dir=/home/TestData/asr/longform/earnings22 \
+        output_filename=preds.json
+      AFTER_SCRIPT: |
+        rm -rf preds.json
+
+  # L2: Longform with TimeStamps from Audio Dir
+  L2_Longform_Speech_Transcription_with_TimeStamps_Canary_Chunked_Infer_from_Audio_Dir:
```
[the remark this answers is not in the document]
Comment (krishnacpuvvada) [reply .github/workflows/cicd-main.yml:668]: they cover different aspects of script. let keep them to have maximum coverage.

**Review by nithinraok — APPROVED** [review]
LGTM. Thanks. I think we can make merging timestamps much simpler by adding offset to process_aed_timestamps. But can be done later.

**Review by nithinraok — APPROVED** [review]

**Comment by krishnacpuvvada** [comment]
@ko3n1g all tests passed,  can you approve ? Thanks!

File: .github/workflows/cicd-main.yml [hunk 6]
```diff
@@ -649,6 +649,52 @@ jobs:
...
+    needs: [pre-flight, cicd-test-container-build]
+    uses: ./.github/workflows/_test_template.yml
+    if: contains(fromJSON(needs.pre-flight.outputs.test_to_run), 'L2_Longform_Speech_Transcription_Canary_Chunked_Infer_from_Audio_Dir')
+    with:
+      RUNNER: self-hosted-azure
+      SCRIPT: |
+        python examples/asr/asr_chunked_inference/aed/speech_to_text_aed_chunked_infer.py \
+        model_path=/home/TestData/asr/canary/models/canary-1b-flash_HF_20250318.nemo \
+        audio_dir=/home/TestData/asr/longform/earnings22 \
+        output_filename=preds.json
```
Comment (ko3n1g) [inline .github/workflows/cicd-main.yml:663]: We have a new format where this goes into dedicated test files. Could refactor that in this or a follow up PR?

**Review by ko3n1g — APPROVED** [review]

File: .github/workflows/cicd-main.yml (hunk 6 above)
Comment (krishnacpuvvada) [reply .github/workflows/cicd-main.yml:663]: planning for a model release asap. Will do a followup PR.
